### ISSUE TITLE
Using XCUITest according to appium documentation

### ISSUE THREAD

**Issue by oscheller1** [appium/appium#8093]
There is an awesome, comprehensive guide of all the different XCUI functionality here: https://appium.io/slate/en/master/?python#querying-for-elements-xcuielementquery

Reading through those makes me want to use many of them, such as `childrenMatchingType`, `containingPredicate`, ...

Is there a way to use the documented functionality from within a Appium client? To my current knowledge, the only option we have using XCUITest directly is something like this: 

```
driver.find_element_by_ios_predicate_string('label == "OK" AND type == "XCUIElementTypeButton"')
``` 

This means we can't do nested searches using predicates and instead must use XPath as soon as we want to have nested queries.

Long story short, is there a way to use all the functionality provided in the documentation mentioned above? And if so, can you give samples on how to use them?

I'm curious if something like this would be possible in the future:

```
driver.find_element_by_xcui(
element('name == "some_id_here" AND type == "XCUIElementTypeTableCell")
.containingPredicate('label == "OK" AND type == "XCUIElementTypeButton"')
)
```

**Comment by mykola-mokhnach** [comment]
We've recently added [class chain](https://github.com/appium/java-client/pull/599) for such purpose (search of nested element). Not sure, though, whether it is supported in python client.

**Comment by oscheller1** [comment]
That looks promising. Is there a documentation about this new locator strategy somewhere? Would it be possible to add additional filters on the different classes used in the locator string? i.e. would it be possible to change `XCUIElementTypeWindow/*/XCUIElementTypeTextField[2]` to something like `XCUIElementTypeWindow/*/XCUIElementTypeTextField[@label == 'sometext']` (i.e. using predicate filters), instead of fixed values in the hierarchy (`[2]`).

**Comment by oscheller1** [comment]
I just took a look at the documentation of the new class chain locator here https://github.com/facebook/WebDriverAgent/wiki/Queries. Unfortunately, 
> This query type is similar to xpath, but can only include indexes and valid class names. Only search by direct children of the current element is supported

So we won't be able to nest multiple predicate functions using this new locator.

**Comment by mykola-mokhnach** [comment]
I'll think about adding such features there - it might be useful, but will require some time. For now you can perform lookup by splitting the stuff into separate steps.

**Comment by oscheller1** [comment]
That sounds good. In fact, anything that can find nested queries and supports descendants/children/parents and filtering using the native attributes (name,label, ...) AND is faster than Xpath is very much appreciated. 

In fact, would it be possible to support native XCUITest string support? On Android we can simply send whole UIAutomator strings to the Appium server. These actions are executed extremely fast and they are also just as mighty as xpath when it comes to nesting queries.

In a perfect world, I'd love using the same pattern in iOS using the XCUI functionality showcased in the documentation https://appium.io/slate/en/master/?python#querying-for-elements-xcuielementquery

**Comment by mykola-mokhnach** [comment]
even if there is such possibility then it means we need to pass raw ObjectiveC code as a string and execute it, which would be super insecure. Although, you're welcome to create your own implementation - it's open source.

**Comment by mykola-mokhnach** [comment]
Adding predicates support looks challenging and I like challenges: https://github.com/facebook/WebDriverAgent/pull/522
:)

**Comment by oscheller1** [comment]
Question regarding the class chain locator. Is searching for children instead of descendants explicitly required to get fast execution times? Could an option be added to search for descendants instead of children or would this be just as slow as xpath locator?

**Comment by mykola-mokhnach** [comment]
Yes, it will be slower then. It's how XCTest is designed.

**Comment by oscheller1** [comment]
I've just done a little research on native usage of descendant vs. children. The conclusion seems to be that using descendant in xcode natively is not less performant than children (if so, it's in the one-digit millisecond area). For this we created two ui tests in xcode, one using decendants (Sample 1), the other one using children explicitly (Sample 2). This is not me saying you're wrong but instead wondering why in native UI tests, this is really performant and comparable (in speed) to Appuium using UIAutomator tests in Android.

## Sample 1 using descendant:
### Test code:
```
XCUIApplication *app = [[XCUIApplication alloc] init];
XCUIElementQuery *tablesQuery = app.tables;
[tablesQuery.staticTexts[@"Debugging"] tap];
[[tablesQuery.cells[@"settings_debugging_backend_connection_select_backend_type"].buttons elementBoundByIndex:0] tap];
```

### Log output:
```
...
    t =    11.53s     Tap "Debugging" StaticText
    t =    11.53s         Wait for app to idle
    t =    11.74s         Find the "Debugging" StaticText
    t =    11.74s             Snapshot accessibility hierarchy for com.bshg.hc.trunk
    t =    11.83s             Find: Descendants matching type Table
    t =    11.84s             Find: Descendants matching type StaticText
    t =    11.84s             Find: Elements matching predicate '"Debugging" IN identifiers'
    t =    11.95s             Wait for app to idle
    t =    12.16s         Synthesize event
    t =    12.35s         Wait for app to idle
    t =    12.90s     Tap Button
    t =    12.90s         Wait for app to idle
    t =    13.13s         Find the Button
    t =    13.13s             Snapshot accessibility hierarchy for com.bshg.hc.trunk
    t =    13.47s             Find: Descendants matching type Table
    t =    13.47s             Find: Descendants matching type Cell
    t =    13.48s             Find: Elements matching predicate '"settings_debugging_backend_connection_select_backend_type" IN identifiers'
    t =    13.48s             Find: Descendants matching type Button
    t =    13.48s             Find: Element at index 0
    t =    13.59s             Wait for app to idle
    t =    13.81s         Synthesize event
    t =    14.01s         Wait for app to idle
```
Time passed for these two taps: 2,48 sec

## Sample 2 using children
```
...
    t =    11.48s     Tap "Debugging" StaticText
    t =    11.48s         Wait for app to idle
    t =    11.69s         Find the "Debugging" StaticText
    t =    11.69s             Snapshot accessibility hierarchy for com.app1
    t =    11.78s             Find: Descendants matching type Table
    t =    11.78s             Find: Descendants matching type StaticText
    t =    11.78s             Find: Elements matching predicate '"Debugging" IN identifiers'
    t =    11.89s             Wait for app to idle
    t =    12.11s         Synthesize event
    t =    12.30s         Wait for app to idle
    t =    12.84s     Tap Button
    t =    12.84s         Wait for app to idle
    t =    13.07s         Find the Button
    t =    13.07s             Snapshot accessibility hierarchy for com.app1
    t =    13.41s             Find: Descendants matching type Table
    t =    13.41s             Find: Children matching type Cell
    t =    13.41s             Find: Elements matching predicate 'identifier == "settings_debugging_backend_connection_select_backend_type"'
    t =    13.41s             Find: Children matching type Button
    t =    13.41s             Find: Element at index 0
    t =    13.53s             Wait for app to idle
    t =    13.75s         Synthesize event
    t =    13.95s         Wait for app to idle
```
Time passed for these two taps: 2,47 sec

In this example we had to use descendant for finding the `table` because it's not a direct child of `app`

## Conclusion
Overall natively running XCUITests is pretty fast when executed directly from xcode. The standard notation (such as `[[tablesQuery.cells[@"settings_debugging_backend_connection_select_backend_type"].buttons elementBoundByIndex:0] tap];`), recorded by the ui automation recorder of xcode uses the `.`-notation which is translated to descendants commands (see logs).
Technically, using descendents rather than children is way more reliable because of potential future changes to the app that might add or remove invisible layers in the UI. Hence it's so important to us to use descendant instead of children.

**Comment by mykola-mokhnach** [comment]
Thanks for the investigation. But how about nested queries - why such strategy won't work in this case?

**Comment by oscheller1** [comment]
Our app is pretty complex. There are a lot of UI layers whose purpose is just to manage the UI appropriately. (Different status bars, menues, navigations, containers)

### Example
A login email field has the following Xpath:
```
//XCUIElementTypeApplication[1]/XCUIElementTypeWindow[1]/
XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/
XCUIElementTypeScrollView[1]/XCUIElementTypeOther[1]/
XCUIElementTypeTextField[1]
```
Querying the edit field by just children from the `XCUIElementTypeWindow` on is no option for us because there is an uncertain/inconsistent number of `XCUIElementTypeOther` following. Instead just searching for an element with the predicate `name == "E-Mail"` will be sufficient for us at this point to find the edit field.

Now, sometimes the element we want to access is not uniquely identifiable. Let's take the button from the example from my last post.
`[[tablesQuery.cells[@"settings_debugging_backend_connection_select_backend_type"].buttons elementBoundByIndex:0] tap];`
The button itsself has no id or label. Instead, one of its ancestors (in this case a table cell) is unique by id. We know that there's only one button inside this unique container, so we just return the 1 button.

Xpath of the button:
```
//XCUIElementTypeApplication[1]/XCUIElementTypeWindow[1]/
XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/
XCUIElementTypeOther[2]/XCUIElementTypeOther[1]/
XCUIElementTypeOther[1]/XCUIElementTypeOther[2]/
XCUIElementTypeOther[1]/XCUIElementTypeTable[1]/
XCUIElementTypeCell[4]/XCUIElementTypeButton[1]
```

In my example, the cell is actually the parent (instead of an ancestor) of the button, but this sometimes might not be the case (even though after checking, most of the time we have direct parents insted of ancestors).

The main problem with only being able to find children of `XCUIElementTypeWindow` (as documented here https://github.com/facebook/WebDriverAgent/pull/522/files XCUIElement+FBClassChain.h line 30+) is the uncertain amount of `XCUIElementTypeOther`'s before the real elements show up, that we actually want to access. 

Using Xpath functionality `//` is currently our best approach when searching for elements that are  uniquely identifiable only when nested with a parent (such as the button from the example). For elements that are unique by themselves, we're using `find_element_by_ios_predicate()` which is sufficient and has a far better performance than xpath.

**Comment by oscheller1** [comment]
When searching for `driver.findElement(By.className("XCUIElementTypeTable"))` I might get the first of 3 different available tables since table by itsself might not be a unqiue query. I would then have to iterate over those tables to find the correct one which includes the element i'm looking for. That would take even longer than searching by xpath.

I've already tried search by nesting WebDriver elements. In the example above, it will just take the first table it finds and nest the next query. If the first table isn't the one containing my element, the query will just fail, stating there is no such element, ignoring the next 2 available tables. Nesting elements that way only works properly if every single element in the nesting list is uniquely identifiable within the given scope.

Here the first query might return the wrong table,
`driver.findElement(By.className("XCUIElementTypeTable")).findElement(By.classChain("*/XCUIElementTypeButton[2]"))`
where as here, the correct element (also the correct table) is found 'automatically':
`driver.find_element_by_xpath("//XCUIElementTypeTable/descendant::XCUIElementTypeButton")`

I'm aware that my problems might be pretty special and I don't expect you to implement new functionality for me. Most of our problems can be fixed by refactoring the iOS app heavily and create unique identifiers for every single element that we want to automate. I'm currently investigating possibilites to avoid this refactoring as that would be a huge undertaking.

**Comment by Brian-Watson** [comment]
@oscheller1 The iOS app I test sounds very very much like yours!  I often have to come up with really "creative" xpath queries to find things correctly, as the accessibilityID support is lacking.  I'll be keeping an eye on this one closely as my iOS test run times are much longer then our Android run times for the exact same tests with an equivalent flowing UI.

I'm definitely going to start experimenting with using predicates instead of xpath for some locators and see how that works now that I have finally converted my iOS tests over to Appium 1.6 (UIAutomation) from Appium 1.5 (XCUITest).

**Comment by vikramvi** [comment]
@Brian-Watson can you please clarify how do predicates look like or how to use it etc, this will be useful for someone who will visit this post for similar problem.

**Comment by mykola-mokhnach** [comment]
@oscheller1 I have prepared an update to class chain locator, which supports indirect search. You can try the patch yourself at https://github.com/facebook/WebDriverAgent/pull/612 
I'm opened to all suggestions.

Don't expect the patch to be merged into master soon, since all WDA reviewers are on vacation.

**Comment by mykola-mokhnach** [comment]
@vikramvi Have you checked [this](https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/ios_predicate.md) document?
Also, there are some examples in WDA [wiki](https://github.com/facebook/WebDriverAgent/wiki/Queries).

**Comment by imurchie** [comment]
Closing due to inactivity.

**Comment by mike-pt** [comment]
Sorry to comment on a closed issue, but I think I'm facing a similar situation...

Consider this block:
```xml
</XCUIElementTypeCell>
<XCUIElementTypeCell type="XCUIElementTypeCell" enabled="true" visible="true" x="0" y="368" width="375" height="175">
    <XCUIElementTypeOther type="XCUIElementTypeOther" name="cell.ElementA.cView.cell" label="Example Label A" enabled="true" visible="true" x="10" y="423" width="142" height="118"/>
    <XCUIElementTypeOther type="XCUIElementTypeOther" name="cell.ElementA.cView.cell" label="Example Label B" enabled="true" visible="true" x="162" y="423" width="142" height="118"/>
    <XCUIElementTypeOther type="XCUIElementTypeOther" name="cell.ElementA.cView.cell" label="Example Label C" enabled="true" visible="true" x="314" y="423" width="142" height="118"/>
    <XCUIElementTypeStaticText type="XCUIElementTypeStaticText" value="Example" name="item.title" label="Example" enabled="true" visible="true" x="10" y="378" width="316" height="45"/>
    <XCUIElementTypeButton type="XCUIElementTypeButton" name="item.item-menu" label="item Menu" enabled="true" visible="true" x="331" y="378" width="44" height="45"/>
</XCUIElementTypeCell>
````

I was using `elem = driver.find_element_by_ios_predicate('wdName = "item.title" AND wdValue = "%s"' % itemName)` to get to the correct item, but I want to click the button...

My frist guess is that I would have to select the *parent* of `elem` and then the XCUIElementTypeButton... is this possible with `find_element_by_ios_class_chain`? I don't see anything about parents just children here: https://github.com/facebook/WebDriverAgent/wiki/Class-Chain-Queries-Construction-Rules

**Comment by mykola-mokhnach** [comment]
@mike-pt you could use `includes` query, like `**/XCUIElementTypeCell[$name BEGINSWITH "D"$]`. This will select a cell(s), which CONTAINS a descendant element whose name starts with D.

**Comment by mike-pt** [comment]
I actually missed the "descendants" part on first read -.-'

I was now trying something like:
```
driver.find_element_by_ios_class_chain('**/XCUIElementTypeCell/XCUIElementTypeStaticText[`name == "item.title" AND value == "Exmaple"`]/**/XCUIElementTypeButton')`
```

But that doesn't find an element (button), this works for the Cell though.
```
driver.find_element_by_ios_class_chain('**/XCUIElementTypeCell/XCUIElementTypeStaticText[`name == "item.title" AND value == "Exmaple"`]')
```

From the docs I got the idea that /**/ would do the search on the descendants, should it should find it right?

**Comment by mike-pt** [comment]
@mykola-mokhnach thanks for the help (forgot to mention before)

So this works:

`**/XCUIElementTypeCell[$value == "Example"$]/XCUIElementTypeButton`

Yet this doesn't
`**/XCUIElementTypeCell[$type == XCUIElementTypeStaticText AND value == "Example"$]/XCUIElementTypeButton` 


Can that be a bug? the value works for me, but according to the doc we should be able to use "AND" and search for type here

**Comment by mykola-mokhnach** [comment]
Type name should be quoted since it is a string

**Comment by mike-pt** [comment]
It is quoted so is the value ... should `XCUIElementTypeStaticText` also be quoted? the examples state different 


EX:
`XCUIElementTypeWindow/XCUIElementTypeTable/XCUIElementTypeCell[`visible == 1`][$type == XCUIElementTypeImage AND name == 'bla'$]/XCUIElementTypeTextField` - select a text field, which is a direct child of a visible table cell, which has at least one descendant image with identifier 'bla'

**Comment by mykola-mokhnach** [comment]
It's a typo. I cannot fix it, because FB restricted the access
